cmd/pj-rehearse/main.go

```
@@ -161,17 +161,32 @@ func main() {
 			gracefulExit(o.noFail, "")
 		}
 	}
+
 	loggers := rehearse.Loggers{Job: logger, Debug: debugLogger.WithField(prowgithub.PrLogField, prNumber)}
-	success, err := rehearse.ExecuteJobs(changedPresubmits, prNumber, o.candidatePath, jobSpec.Refs, !o.dryRun, loggers, pjclient)
-	if err != nil {
-		logger.WithError(err).Error("Failed to rehearse jobs")
-		gracefulExit(o.noFail, rehearseFailureOutput)
-	}
-	if !success {
-		logger.Error("Some jobs failed their rehearsal runs")
-		gracefulExit(o.noFail, jobsFailureOutput)
+
+	resultChan := make(chan rehearse.Result)
+	errorChan := make(chan error)
+	quit := make(chan struct{})
+
+	executor := rehearse.NewExecutor(changedPresubmits, prNumber, o.candidatePath,
+		jobSpec.Refs, loggers, pjclient, o.dryRun, resultChan, errorChan, quit)
+
+	go executor.ExecuteJobs()
+
+	results := make(map[string]pjapi.ProwJobStatus)
+	for {
+		select {
+		case result := <-resultChan:
+			results[result.Name] = result.Status
+		case err := <-errorChan:
+			loggers.Job.WithError(err).Error("Error has occurred")
```

The failure is not propagated to the end result.